Multi-fidelity Bayesian optimization notebook

marpulli opened the pull request:
Multi-fidelity Bayesian optimization notebook + multi information source entropy search acquisition function

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

javiergonzalezh commented on emukit/model_wrappers/gpy_model_wrappers.py:
```diff
@@ -96,12 +96,14 @@ def Y(self) -> np.ndarray:
 
 
-class GPyMultiOutputWrapper(IModel, IDifferentiable, ICalculateVarianceReduction):
+class GPyMultiOutputWrapper(IModel, IDifferentiable, ICalculateVarianceReduction, IEntropySearchModel):
```
What if I want to use a multiple-output GP with an acquisition that is not ES. Would this require to always have the ES interface?

marpulli replied:
I don't fully understand the comment but this change allows you to use entropy search but does not limit you doing anything else.

javiergonzalezh replied:
OK, fine, it is GPy specific so no issues.

mmahsereci commented:
Could you include the BO high-fidelity-only baseline in the notebook? I find it less convincing without it since it does not showcase why one would want to do MF instead of vanilla BO.

javiergonzalezh commented on notebooks/Emukit-multi-fidelity-bayesian-optimization.ipynb:
```diff
@@ -0,0 +1,564 @@
+   "outputs": [
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAvQAAAH/CAYAAAA480lIAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4wLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvqOYd8AAAIABJREFUeJzs3XmcjXX/x/HX1xjDGDtjZ+yUkEYoMboxyRJC3SGy0yJ3m+7EFEn77S6FSkoLv4QWJbqtZSqS7FOy7/vOmOX7++M6Z8xyZsyYM3PMzPv5eJzHmXOtn+s6M3yu7/X5fi9jrUVERERERHKmfL4OQERERERErp4SehERERGRHEwJvYiIiIhIDqa ...
```
Let's also cite this one here
https://papers.nips.cc/paper/5086-multi-task-bayesian-optimization.pdf

javiergonzalezh commented on emukit/bayesian_optimization/acquisitions/entropy_search.py:
```diff
@@ -208,3 +215,70 @@ def _innovations(self, x: np.ndarray) -> tuple:
+
+        # Find information source parameter in parameter space
+        info_source_parameter = None
+        self.fidelity_idx = None
```
Is it right to call this fidelity if this class is general for muti output/fidelity cases?

javiergonzalezh commented on emukit/bayesian_optimization/acquisitions/entropy_search.py:
```diff
@@ -208,3 +215,70 @@ def _innovations(self, x: np.ndarray) -> tuple:
+        for i, param in enumerate(space.parameters):
+            if isinstance(param, InformationSourceParameter):
+                info_source_parameter = param
+                self.fidelity_idx = i
```
same here, shall we use fidelity?

marpulli commented:
@mmahsereci I'll add a high fidelity only baseline at the end of the notebook

marpulli commented:
- Added a comparison to high fidelity only at the end of the notebook.
- Changed fidelity -> index in entropy search code
- Added Multi-Task Bayes opt reference to notebook

apaleyes commented on emukit/bayesian_optimization/acquisitions/entropy_search.py:
```diff
@@ -214,3 +215,70 @@ def _innovations(self, x: np.ndarray) -> tuple:
+    Entropy search acquisition for multi-information source problems where the objective function is the output of one
+    of the information sources. The other information sources provide auxiliary information about the objective function
+    """
+    def __init__(self, model: Union[IModel, IEntropySearchModel], space: ParameterSpace,
```
Any chance you could move part of this __init__ in a separate method, to shorten it a bit? Maybe identifying info source param piece? Maybe creation of proposal function?

marpulli replied:
Done

apaleyes commented on emukit/bayesian_optimization/acquisitions/entropy_search.py:
```diff
@@ -214,3 +215,70 @@ def _innovations(self, x: np.ndarray) -> tuple:
         return False
+
+
+class MultiInformationSourceEntropySearch(EntropySearch):
```
The fact that we can reuse stuff like that is just great!

apaleyes commented on notebooks/Emukit-multifidelity-tutorial.ipynb:
```diff
@@ -847,6 +847,13 @@
     "- Perdikaris, P., Raissi, M., Damianou, A., Lawrence, N.D. and Karniadakis, G.E., 2017. *Nonlinear information fusion algorithms for data-efficient multi-fidelity modelling.* Proc. R. Soc. A, 473(2198), p.20160751."
    ]
+  },
+  {
```
Empty cell?

marpulli replied:
Removed this file from PR

apaleyes commented on tests/emukit/bayesian_optimization/test_entropy_search.py:
```diff
@@ -55,3 +66,12 @@ def test_innovations(acquisition):
+def test_multi_information_source_entropy_search_shape(multi_source_acquisition):
+    rng = np.random.RandomState(43)
+
+    x_test = np.concatenate([rng.rand(2, 1), np.array([[0], [1]])], axis=1)
```
Fixing random seed and calling random is just as good as hardcoding numbers. The latter however is simpler.

That is, unless random is also called somewhere down the line, but does not look like there is any

marpulli replied:
Done

apaleyes commented on notebooks/Emukit-multi-fidelity-bayesian-optimization.ipynb:
```diff
@@ -0,0 +1,692 @@
+    "\n",
+    "Both are defined in the interval [0, 1].\n",
+    "\n",
+    "The high fidelity function is assigned a cost of $3$, whereas the low fidelity function is assigned a cost of $1$.\n",
```
"is assigned a cost of X" is a bit unclear. "Each single evaluation of f costs X" would be clearer to me

marpulli replied:
Changed

apaleyes replied:
Is it?

marpulli replied:
Weird, I'm sure I changed them in the github editor. I'll try again

apaleyes commented on notebooks/Emukit-multi-fidelity-bayesian-optimization.ipynb:
```diff
@@ -0,0 +1,692 @@
+   "source": [
+    "### Define Model\n",
+    "\n",
+    "We will use the linear multi-fidelity model defiend in Emukit. In this model, the high-fidelity function is modelled as a scaled sum of the low-fidelity function plus an error term:\n",
```
```suggestion
    "We will use the linear multi-fidelity model defined in Emukit. In this model, the high-fidelity function is modelled as a scaled sum of the low-fidelity function plus an error term:\n",
```

apaleyes commented on notebooks/Emukit-multi-fidelity-bayesian-optimization.ipynb:
```diff
@@ -0,0 +1,692 @@
+   "outputs": [
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAvQAAAH/CAYAAAA480lIAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4wLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvqOYd8AAAIABJREFUeJzs3Xd4lFXax/HvSYAUQid0MHRsgBJXUKoCoSgrKFhQQUCaBV37WogNUbGt4gLrIi6iYAFXdKX40ixBQTpSRHrvhBpI5rx/nElImUBCQiYz/D7XNdeQp97zTKL3c577nGOstYiIiIiISGAK8XcAIiIiIiJy7pTQi4iIiIgEMCX0IiIiIiIBTAm ...
```
A sentence about why and how we subscribe to an event to make that happen would be helpful

marpulli replied:
Done

apaleyes commented on notebooks/Emukit-multi-fidelity-bayesian-optimization.ipynb:
```diff
@@ -0,0 +1,692 @@
+   "outputs": [
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAvQAAAH/CAYAAAA480lIAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4wLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvqOYd8AAAIABJREFUeJzs3Xd4lFXax/HvSYAUQid0MHRsgBJXUKoCoSgrKFhQQUCaBV37WogNUbGt4gLrIi6iYAFXdKX40ixBQTpSRHrvhBpI5rx/nElImUBCQiYz/D7XNdeQp97zTKL3c577nGOstYiIiIiISGAK8XcAIiIiIiJy7pTQi4iIiIgEMCX0IiIiIiIBTAm ...
```
Very nice to see events in action

apaleyes commented on notebooks/Emukit-multi-fidelity-bayesian-optimization.ipynb:
```diff
@@ -0,0 +1,692 @@
+   "outputs": [
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAvQAAAH/CAYAAAA480lIAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4wLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvqOYd8AAAIABJREFUeJzs3Xd4lFXax/HvSYAUQid0MHRsgBJXUKoCoSgrKFhQQUCaBV37WogNUbGt4gLrIi6iYAFXdKX40ixBQTpSRHrvhBpI5rx/nElImUBCQiYz/D7XNdeQp97zTKL3c577nGOstYiIiIiISGAK8XcAIiIiIiJy7pTQi4iIiIgEMCX0IiIiIiIBTAm ...
```
This sentence lacks a verb somehow

marpulli replied:
Fixed

apaleyes commented on notebooks/Emukit-multi-fidelity-bayesian-optimization.ipynb:
```diff
@@ -0,0 +1,692 @@
+   "outputs": [
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAvQAAAH/CAYAAAA480lIAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4wLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvqOYd8AAAIABJREFUeJzs3Xd4lFXax/HvSYAUQid0MHRsgBJXUKoCoSgrKFhQQUCaBV37WogNUbGt4gLrIi6iYAFXdKX40ixBQTpSRHrvhBpI5rx/nElImUBCQiYz/D7XNdeQp97zTKL3c577nGOstYiIiIiISGAK8XcAIiIiIiJy7pTQi4iIiIgEMCX0IiIiIiIBTAm ...
```
A final sentence analyzing the comparison figure would be nice.

marpulli replied:
Added

apaleyes commented on emukit/bayesian_optimization/acquisitions/entropy_search.py:
```diff
@@ -214,3 +215,85 @@ def _innovations(self, x: np.ndarray) -> tuple:
+        """
+
+        # Find information source parameter in parameter space
+        info_source_parameter, source_idx = MultiInformationSourceEntropySearch._find_source_parameter(space)
```
Underscore normally means private member, so calling it like that would be highly not advisable. Make "find_source_parameter" a package level function and use it like that
